teps/0094-configuring-resources-at-runtime.md

```diff
@@ -52,12 +61,495 @@ While this may be a valuable feature, it should be considered in a separate prop
+`Task.Step.Resources`, `Task.StepTemplate.Resources`, and `TaskRun.StepResources`
+takes precedence (see [Merging resource requirements](#merging-resource-requirements)).
+In addition, `StepResources` and `SidecarResources` may be confused with `PipelineResource`s.
+This risk can be mitigated via documentation.
+
+### User Experience
+
+The current workaround for lack of this feature is to write a new `Task`
+for each set of resource constraints, as described in
+[this comment](https://github.com/tektoncd/pipeline/issues/4080#issuecomment-884958486)
+from a buildah user.
+
+This proposal moves environment-specific configuration into `TaskRun` definitions,
+allowing `Task` definitions to be reused.
+We may want to add an option to the CLI to specify resource requirements when starting
+a `TaskRun` via `tkn task start`, but this is not necessary for the initial implementation.
+
+## Test Plan
+
+Unit tests should suffice for this feature, covering the following cases:
+- `Task`s with no resource requirements specified
+- `Task`s with resource requirements that are partially overridden by the `TaskRun`
+- `Task`s with resource requirements that are fully overridden by the `TaskRun`
+- `TaskRun`s with resource requirements launched by themselves and from `PipelineRun`s.
+
+Examples should be included for overriding resource requirements
+(e.g. the example used in [Merging resource requirements](#merging-resource-requirements)).
+
+## Design Evaluation
+### Reusability
+This proposal increases reusability of `Task`s and `Pipeline`s by allowing
+environment-specific execution requirements to be updated at runtime.
+
+### Simplicity
+The proposed solution contains the minimum number of features that meet the specified
+requirements, compared to the listed alternatives. 
+
+### Flexibility
+This proposal increases Tekton's flexibility by giving users more options to modify `Task`s.
+There isn't a clear strategy for implementing this functionality via a plugin system.
+
+### Conformance
+Tekton aims to minimize Kubernetes-specific features in its API.
+However, the usage of `ResourceRequirements` is necessary for this feature, as a result of the
+decision to directly embed `Container` in the `Task` API.
+
+Container resource requirements are required for
+[Knative Serving conformance](https://github.com/knative/specs/blob/main/specs/serving/knative-api-specification-1.0.md#container)
+but not for
+[Tekton pipelines conformance](https://github.com/tektoncd/pipeline/blob/main/docs/api-spec.md#step).
+Therefore, `StepResources` and `SidecarResources` should also not be required for Tekton conformance.
+
+## Drawbacks
+
+In an ideal world, `Task`s would not contain fields that are tied to runtime requirements.
+`Task`s might be more reusable if `Step` and `Sidecar` were fully Tekton-owned,
+instead of having the `Container` API embedded.
+However, this would be a major API change at this point,
+for little gain compared to the proposed solution.
```

Comment:
is it possible to elaborate a bit on the "little gain"?

Reply:
nit, just spit balling

im also not clear on whether this change would solve the problem completely - maybe it makes sense in some cases to specify resource limits at authoring time as well and you'd still need to override them?

e.g. esp for scenarios within a particular organization, it might make more sense to be able to assume certain resource constraints

Reply:
Added a few thoughts but I don't have many specific examples. It seems like resource requirements is the primary way this causes friction at the moment.